flink-java/src/main/java/org/apache/flink/api/java/sampling/BernoulliSampler.java

```diff
@@ -102,15 +106,31 @@ public T next() {
 			}
 
 			private T getNextSampledElement() {
-				while (input.hasNext()) {
-					T element = input.next();
-
-					if (random.nextDouble() <= fraction) {
+				if (fraction <= THRESHOLD) {
+					double rand = random.nextDouble();
+					double u = Math.max(rand, EPSILON);
+					int gap = (int) (Math.log(u) / Math.log(1 - fraction));
+					int elementCount = 0;
+					if (input.hasNext()) {
+						T element = input.next();
+						while (input.hasNext() && elementCount < gap) {
```

Comment:
While gap is larger than the input remain elements number, this should return null, right? but the implementation here return the last input element.